overlord/ifacestate/hotplug.go

```diff
@@ -0,0 +1,248 @@
+		}
+
+		chg := st.NewChange(fmt.Sprintf("hotplug-connect-%s", iface), fmt.Sprintf("Connect hotplug slots of interface %s", iface))
+		hotplugConnect := st.NewTask("hotplug-connect", fmt.Sprintf("Recreate connections of device %q", key))
+		hotplugTaskSetAttrs(hotplugConnect, key, iface.Name())
+		chg.AddTask(hotplugConnect)
+		st.EnsureBefore(0)
+	}
+}
+
+// HotplugDeviceRemoved gets called when a device is removed from the system.
+func (m *InterfaceManager) HotplugDeviceRemoved(devinfo *hotplug.HotplugDeviceInfo) {
+	st := m.state
+	st.Lock()
+	defer st.Unlock()
+
+	hotplugFeature, err := m.hotplugEnabled()
+	if err != nil {
+		logger.Noticef(err.Error())
+		return
+	}
+
+	defaultKey := defaultDeviceKey(devinfo)
+	for _, iface := range m.repo.AllHotplugInterfaces() {
+		// determine device key for the interface; note that interface might provide own device keys.
+		key, err := deviceKey(defaultKey, devinfo, iface)
+		if err != nil {
+			logger.Debugf(err.Error())
+			continue
+		}
+
+		if !validateDeviceKey(key) {
+			continue
+		}
+
+		hasSlots, err := m.repo.HasHotplugSlots(key, iface.Name())
+		if err != nil {
+			logger.Noticef("Cannot check hotplug slots: %s", err)
+			continue
+		}
+		if !hasSlots {
+			continue
+		}
+
+		if !hotplugFeature {
+			logger.Noticef("Hotplug 'remove' event for device %q (interface %q) ignored, enable experimental.hotplug", devinfo.DevicePath(), iface.Name())
+			continue
+		}
+
+		logger.Debugf("HotplugDeviceRemoved: %s (default key: %q, devname %s, subsystem: %s)", devinfo.DevicePath(), defaultKey, devinfo.DeviceName(), devinfo.Subsystem())
+
+		// create tasks to disconnect given interface
+		chg := st.NewChange(fmt.Sprintf("hotplug-remove-%s", iface), fmt.Sprintf("Remove hotplug connections and slots of interface %s", iface))
+
+		// hotplug-disconnect task will create hooks and disconnect all the slots
+		hotplugRemove := st.NewTask("hotplug-disconnect", fmt.Sprintf("Disable connections of device %q", key))
+		hotplugTaskSetAttrs(hotplugRemove, key, iface.Name())
+		chg.AddTask(hotplugRemove)
+
+		// hotplug-remove-slot will remove this device's slots from the repository.
+		removeSlots := st.NewTask("hotplug-remove-slots", fmt.Sprintf("Remove slots for device %q, interface %q", key, iface))
+		hotplugTaskSetAttrs(removeSlots, key, iface.Name())
+		removeSlots.WaitFor(hotplugRemove)
+		chg.AddTask(removeSlots)
+		st.EnsureBefore(0)
+	}
+}
+
+func findConnsForDeviceKey(conns *map[string]connState, coreSnapName, ifaceName, deviceKey string) []string {
+	var connsForDevice []string
+	for id, connSt := range *conns {
+		if connSt.Interface != ifaceName || connSt.HotplugDeviceKey != deviceKey {
+			continue
+		}
+		connsForDevice = append(connsForDevice, id)
+	}
+	return connsForDevice
+}
+
+func (m *InterfaceManager) hotplugEnabled() (bool, error) {
+	tr := config.NewTransaction(m.state)
+	var featureFlagHotplug bool
+	if err := tr.GetMaybe("core", "experimental.hotplug", &featureFlagHotplug); err != nil {
```

Comment:
`snapstate` has a similar helper `getFeatureFlagBool` that handles a case when the feature was disabled via `snap set system experimenta.feature=` in which case deserialization to `bool` will fail. Maybe you could reuse that code or make it public.

Reply:
Ah indeed, I forgot we had a helper already, thanks! I'll make it public.